pkg/api/api.go

```diff
@@ -0,0 +1,229 @@
+	resp, body, err := client.Do(ctx, req)
+	if err != nil {
+		return resp, body, nil, err
+	}
+
+	code := resp.StatusCode
+
+	if code/100 != 2 && !apiError(code) {
+		errorType, errorMsg := errorTypeAndMsgFor(resp)
+		return resp, body, nil, &promapiv1.Error{
+			Type:   errorType,
+			Msg:    errorMsg,
+			Detail: string(body),
+		}
+	}
+
+	var result apiResponse
+
+	if http.StatusNoContent != code {
+		if jsonErr := json.Unmarshal(body, &result); jsonErr != nil {
+			return resp, body, nil, &promapiv1.Error{
+				Type: ErrBadResponse,
+				Msg:  jsonErr.Error(),
+			}
+		}
+	}
+
+	if apiError(code) && result.Status == "success" {
+		err = &promapiv1.Error{
+			Type: ErrBadResponse,
+			Msg:  "inconsistent body for response code",
+		}
+	}
+
+	if result.Status == "error" {
+		err = &promapiv1.Error{
+			Type: result.ErrorType,
+			Msg:  result.Error,
+		}
+	}
+
+	return resp, result.Data, result.Warnings, err
+}
+
+// doGetFallback will attempt to do the request as-is, and on a 405 it will fallback to a GET request.
+func doGetFallback(
+	ctx context.Context,
+	client promapi.Client,
+	u *url.URL,
+	args url.Values,
+	cache bool,
+) (*http.Response, []byte, promapiv1.Warnings, error) { //nolint:unparam
+	req, err := http.NewRequest(http.MethodPost, u.String(), strings.NewReader(args.Encode()))
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	if !cache {
+		req.Header.Set("Cache-Control", "no-store")
+	}
+
+	resp, body, warnings, err := do(ctx, client, req)
+	if resp != nil && resp.StatusCode == http.StatusMethodNotAllowed {
+		u.RawQuery = args.Encode()
+		req, err = http.NewRequest(http.MethodGet, u.String(), nil)
+		if err != nil {
+			return nil, nil, warnings, err
+		}
+
+	} else {
+		if err != nil {
+			return resp, body, warnings, err
+		}
+		return resp, body, warnings, nil
+	}
+	return do(ctx, client, req)
+}
+
+func QueryRange(ctx context.Context, client promapi.Client, query string, r promapiv1.Range,
+	cache bool) (model.Value, promapiv1.Warnings, error) {
+	u := client.URL(epQueryRange, nil)
+	q := u.Query()
+
+	q.Set("query", query)
+	q.Set("start", formatTime(r.Start))
+	q.Set("end", formatTime(r.End))
+	q.Set("step", strconv.FormatFloat(r.Step.Seconds(), 'f', -1, 64))
+
+	_, body, warnings, err := doGetFallback(ctx, client, u, q, cache) //nolint:bodyclose
+	if err != nil {
+		return nil, warnings, err
+	}
+
+	var qres queryResult
+
+	return qres.v, warnings, json.Unmarshal(body, &qres)
+}
+
+func Query(ctx context.Context, client promapi.Client, query string, ts time.Time) (model.Value, promapiv1.Warnings, error) {
+	u := client.URL("", nil)
+	q := u.Query()
+
+	q.Set("query", query)
+	if !ts.IsZero() {
+		q.Set("time", formatTime(ts))
+	}
+
+	// Instant query doesn't support cache
+	_, body, warnings, err := doGetFallback(ctx, client, u, q, false) //nolint:bodyclose
```

Comment:
Actually... let's close body! and readAll instead of nolint ;p

See `runutil.ExhaustClose` functions from Thanos

Reply:
Actually it is not the same thing. The body is `[]byte`, I will rename this variable to `data` so that it would be clear.

Reply:
Umm, seems this is a bug of bodyclose linter.

The body is already closed in https://github.com/prometheus/client_golang/blob/master/api/client.go#L121